lib/spack/docs/basic_usage.rst

```diff
@@ -695,6 +695,130 @@ structured the way you want:
     }
 
 
+^^^^^^^^^^^^^^
+``spack diff``
+^^^^^^^^^^^^^^
+
+It's often the case that you have two versions of a spec that you need to
+disambiguate. Let's say that we've installed two variants of zlib, one with
+and one without the optimize variant:
+
+.. code-block:: console
+
+   $ spack install zlib
+   $ spack install zlib -optimize
+
+When we do ``spack find`` we see the two versions.
+
+.. code-block:: console
+
+    $ spack find zlib
+    ==> 2 installed packages
+    -- linux-ubuntu20.04-skylake / gcc@9.3.0 ------------------------
+    zlib@1.2.11  zlib@1.2.11
+
+
+Let's now say that we want to uninstall zlib. We run the command, and hit a problem
+real quickly since we have two!
+
+.. code-block:: console
+
+    $ spack uninstall zlib
+    ==> Error: zlib matches multiple packages:
+
+        -- linux-ubuntu20.04-skylake / gcc@9.3.0 ------------------------
+        efzjziy zlib@1.2.11  sl7m27m zlib@1.2.11
+
+    ==> Error: You can either:
+        a) use a more specific spec, or
+        b) specify the spec by its hash (e.g. `spack uninstall /hash`), or
+        c) use `spack uninstall --all` to uninstall ALL matching specs.
+
+Oh no! We can see from the above that we have two different versions of zlib installed,
+and the only difference between the two is the hash. This is a good use case for 
+``spack diff``, which can easily show us the "diff" or set difference 
+between properties for two packages. Let's try it out.
+Since the only difference between these two is the hash, we provide the hashes:
+
+.. code-block::console
+
+    $ spack diff /efzjziy /sl7m27m
+    ==> diff(zlib@1.2.11/efzjziy, zlib@1.2.11/sl7m27m)
+    VARIANT_SET
+      zlib optimize bool(False)
+    ==> diff(zlib@1.2.11/sl7m27m, zlib@1.2.11/efzjziy)
+    VARIANT_SET
+      zlib optimize bool(True)
+
+
+Awesome! The above tells us that our first zlib was built without optimize (False)
+and the second was built with optimize (True). This is a small example, but there are
+actually several kinds of differences that you can view, a ``VARIANT_SET``
+being just one of them. The first package that you provide (A)
+being diffed against B means that we see what is in A but not B. Here is another example
+with an additional difference type, ``VERSION``:
+
+.. code-block::console
+
+    $ spack diff python@2.7.8 python@3.8.8
+    ==> diff(python@2.7.8/7oknfqf, python@3.8.8/vrp4fmj)
+    VARIANT_SET
+      python patches a8c52415a8b03c0e5f28b5d52ae498f7a7e602007db2b9554df28cd5685839b8
+    VERSION
+      openssl Version(1.0.2u)
+      python Version(2.7.8)
+    ==> diff(python@3.8.8/vrp4fmj, python@2.7.8/7oknfqf)
+    VARIANT_SET
+      python patches 0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87
+    VERSION
+      openssl Version(1.1.1j)
+      python Version(3.8.8)
```

Comment:
Ok I have a slight tweak on this output that I think keeps it simple and (I think) in the spirit of this PR -- can we do something like this? (replacing `foo_set`'s with `foo`):

```diff
$ spack diff python@2.7.8 python@3.8.8
--- python@2.7.8/7oknfqf     # but show full hash here
+++ python@3.8.8/vrp4fmj     # and here
@@ variant @@
-  python patches a8c52415a8b03c0e5f28b5d52ae498f7a7e602007db2b9554df28cd5685839b8
+  python patches 0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87
@@ version @@
-  openssl 1.0.2u
+  openssl 1.1.1j
-  python 2.7.8
+  python 3.8.8
```

The kind of cool thing about this is it's basically `git diff` or `git -u` output and you can format it in an interpretable way (even in GitHub markdown for discussions).  We could add more formats later but I think at least people used to reading GitHub diffs would understand this.  Also we can steal console colors from `diff` like above.

Thoughts?

Reply:
The reason I think we want to keep set is because that's what the facts are output as (or a subset) and that's the only way someone can then provide them as attributes. If we artificially remove set (or do any kind of edit) then a person is going to ask for a specific set of attributes (and not get them) and be really confused. And if we try to still support that, doing so means always doing a conversion between input (without set) --> input__set and then back to --> input.

Reply:
we sorted this out in https://github.com/spack/spack/pull/22283#discussion_r679545826.

Do you like the diff output though?

Reply:
The json is fine (it's json!) and I liked the original output that I opened the PR with back in March, but through review here it was changed to its current form (which I don't think is great) but I'm fine with what others want. I liked the spacing and blue that I had before.

Reply:
okay I added the spacing back, that makes it slightly less terrible than before :laughing: 

If you have a preference for output just tell me and I'll change it. It's just so subjective there is no way to logically choose how it looks.

Reply:
I have OCD about two things here:

1. no capitals (spack isn’t shouty and the names in the output should be what you’d provide on the command line.
2. Unified output (I.e. the individual differences are on consecutive lines like diff does it)

My main reason for wanting something like `diff` is that for large specs the format above will be hard for comparisons — you’ll have to scroll up and down to see the diffs for different sections.

So what if we do your titles (blue) but in lowercase, with actual field values in red/green with either < and > or - and + for left/right?

I don’t really care if the headers have @@ — it just makes the GitHub formatting look nicer bc it’s what `diff -u` does.

Reply:
okay I'll have to refactor the printing so they aren't done separately then.

Reply:
```diff
--- zlib@1.2.11efzjziyc3dmb5h5u5azsthgbgog5mj7g
+++ zlib@1.2.11sl7m27mzkbejtkrajigj3a3m37ygv4u2
@@ Variant_value @@
-  zlib optimize bool(False)
+  zlib optimize bool(True)
```

Reply:
```diff
 bin/spack diff python@2.7.8 python@3.8.11
==> Warning: This interface is subject to change.

--- python@2.7.8tsxdi6gl4lihp25qrm4d6nys3nypufbf
+++ python@3.8.11yjtseru4nbpllbaxb46q7wfkyxbuvzxx
@@ version @@
-  openssl Version(1.0.2u)
+  openssl Version(1.1.1k)
-  python Version(2.7.8)
+  python Version(3.8.11)
@@ variant_value @@
-  python patches a8c52415a8b03c0e5f28b5d52ae498f7a7e602007db2b9554df28cd5685839b8
+  python patches 0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87
```

Reply:
Only nitpick is the hash should have a / before it but this looks awesome!

Reply:
Oh oops, I must have deleted it when I expanded it out.